core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala

```diff
@@ -216,20 +218,27 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
       val uri = new URI(primaryResource)
       val uriScheme = uri.getScheme()
 
-      uriScheme match {
-        case "file" =>
-          try {
+      try {
+        uriScheme match {
+          case "file" =>
             Utils.tryWithResource(new JarFile(uri.getPath)) { jar =>
               // Note that this might still return null if no main-class is set; we catch that later
               mainClass = jar.getManifest.getMainAttributes.getValue("Main-Class")
             }
-          } catch {
-            case _: Exception =>
-              error(s"Cannot load main class from JAR $primaryResource")
-          }
-        case _ =>
+          case _ =>
+            val sparkConf = new SparkConf()
+            sparkConf.setAll(sparkProperties)
+            val fs = FileSystem.get(uri, SparkHadoopUtil.newConfiguration(sparkConf))
+
+            Utils.tryWithResource(new JarInputStream(fs.open(new Path(primaryResource)))) { jar =>
```

Comment:
When submitting in client mode, this will cause a duplicate read of the remote jar in the driver process, because the `SparkSubmit` class also does the download to make it available to the driver.

I think it would be better to move this whole "figure out the main class if it's not defined" logic to `SparkSubmit` after that is done (around L370, look for `var localPrimaryResource: String`).

In cluster mode you'll still need code similar to this to read the manifest, but that should be easy to handle.

Reply:
Thanks, that's really valuable input. I'm not a big fan of my own approach of having Hadoop classes being used in `SparkSubmitArguments` anyway. I'll take a look into this, but probably won't be able to get into it until this weekend.

Reply:
@igozali have you had a chance to take another look at this?

Reply:
Hi @srowen My bad, I dropped the ball here - got caught up in work. Still interested in this though, and will take a look at it before tomorrow night.